### ISSUE TITLE
Build failed when enable squashfs support in uboot

### ISSUE THREAD

**Issue by soccercheng** [foss-for-synopsys-dwc-arc-processors/buildroot#15]
~/arc/prj/out/host/bin/arc-buildroot-linux-gnu-ld.bfd: fs/built-in.o: in function `sqfs_calc_n_blks':
~/arc/prj/out/build/uboot-custom/fs/squashfs/sqfs.c:90: undefined reference to `__umoddi3'
~/arc/prj/out/host/bin/arc-buildroot-linux-gnu-ld.bfd: ~/arc/prj/out/build/uboot-custom/fs/squashfs/sqfs.c:90: undefined reference to `__umoddi3'
~/arc/prj/out/host/bin/arc-buildroot-linux-gnu-ld.bfd: fs/built-in.o: in function `sqfs_read':
~/arc/prj/out/build/uboot-custom/fs/squashfs/sqfs.c:1442: undefined reference to `__umoddi3'
~/arc/prj/out/host/bin/arc-buildroot-linux-gnu-ld.bfd: ~/arc/prj/out/build/uboot-custom/fs/squashfs/sqfs.c:1442: undefined reference to `__umoddi3'
~/arc/prj/out/host/bin/arc-buildroot-linux-gnu-ld.bfd: ~/arc/prj/out/build/uboot-custom/fs/squashfs/sqfs.c:1503: undefined reference to `__umoddi3'
~/arc/prj/out/host/bin/arc-buildroot-linux-gnu-ld.bfd: fs/built-in.o:~/arc/prj/out/build/uboot-custom/fs/squashfs/sqfs.c:1503: more undefined references to `__umoddi3' follow
make[1]: *** [Makefile:1759: u-boot] Error 1
make[1]: Leaving directory '~/arc/prj/out/build/uboot-custom'
make: *** [package/pkg-generic.mk:250: ~/arc/prj/out/build/uboot-custom/.stamp_built] Error 2
make: Leaving directory '~/arc/prj/buildroot-2021.02.2'

**Comment by abrodkin** [comment]
We "imported" a part of `libgcc` into U-Boot to be toolchain independent, but looks like not all the symbols were used in those builds that we used to play with, thus `__umoddi3` is missing. And the simples solution is to not rely on that so-called "private" libs, but use whatever is provided by the toolchain itself. For that just disable `CONFIG_USE_PRIVATE_LIBGCC` option in U-Boot's configuration (i.e. say `make menuconfig` and unset it for HSDK).

Let me know if that helps.

**Comment by soccercheng** [comment]
> We "imported" a part of `libgcc` into U-Boot to be toolchain independent, but looks like not all the symbols were used in those builds that we used to play with, thus `__umoddi3` is missing. And the simples solution is to not rely on that so-called "private" libs, but use whatever is provided by the toolchain itself. For that just disable `CONFIG_USE_PRIVATE_LIBGCC` option in U-Boot's configuration (i.e. say `make menuconfig` and unset it for HSDK).
> 
> Let me know if that helps.

Yes, it worked after disabling CONFIG_USE_PRIVATE_LIBGCC, ---> THANX.

BTW, is CONFIG_USE_PRIVATE_LIBGCC started from "ARC" platform support?

**Comment by abrodkin** [comment]
> BTW, is CONFIG_USE_PRIVATE_LIBGCC started from "ARC" platform support?

Not clear what do you really mean, sorry.
That concept of "private libgcc" existed long before ARC port was introduced I guess. The beauty of "private libgcc" is complete self-contained source tree of U-Boot because no external pre-built libraries (those that otherwise come from the toolchain, typically `libgcc.a`). Otherwise if you rely on pre-built `libgcc.a` from the toolchain you have to be sure that library from your toolchain matches your expectations.

See a compiler (for example GCC) may generate object files (`.o`) for any flavor of the target CPU  regardless [compiler] configuration. I.e. even GCC from "big-endian" toolchain may easily produce "little-endian" object files if you pass `-mlittle-endian` option to it. While `libgcc.a` of that "big-endian" toolchain will be also "big-endian" and you won't be able to build a "little-endian" U-Boot with "big-endian" toolchain if pre-built `libgcc.a` is used.

But if `libgcc.a` (or rather some of its functions) is being built from sources as a part of U-Boot, it's possible to use  ARC GNU toolchain of any endianess as well as optimized for any target ARC processor (i.e. GCC configured woth `--with-cpu=arcem` will build U-Boot for ARC HS perfectly fine).

As an example, with suggested change (disabled "private libgcc") it's no longer possible to use "big-endian" toolchain for building U-Boot for a "little-endian" HSDK, you'll see a lot of such kind of messages during the final linkage stage:
```
...
arceb-elf32-ld.bfd: arch/arc/lib/start.o: compiled for a big endian system and target is little endian
arceb-elf32-ld.bfd: failed to merge target specific data of file arch/arc/lib/start.o
arceb-elf32-ld.bfd: arch/arc/cpu/arcv2/ivt.o: compiled for a big endian system and target is little endian
arceb-elf32-ld.bfd: failed to merge target specific data of file arch/arc/cpu/arcv2/ivt.o
...
```

Still this feature ("private libgcc") is more of a convenience option for us U-Boot developers who need to build it for multiple different platforms now and then. For those who work on a particular product with a fixed IP configuration more traditional approach with prebuilt `libgcc.a` from the toolchain should be good enough.

I hope my explanation makes sense :)
